include/pmacc/math/Matrix.hpp

```diff
@@ -0,0 +1,201 @@
+/* Copyright 2023 Brian Marre, Rene Widera
+ *
+ * This file is part of PMacc.
+ *
+ * PMacc is free software: you can redistribute it and/or modify
+ * it under the terms of either the GNU General Public License or
+ * the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * PMacc is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License and the GNU Lesser General Public License
+ * for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * and the GNU Lesser General Public License along with PMacc.
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+//! @file implementation of matrix and matrix multiplication
+
+#pragma once
+
+#include "pmacc/attribute/unroll.hpp"
+#include "pmacc/math/Vector.hpp"
+#include "pmacc/memory/Array.hpp"
+#include "pmacc/static_assert.hpp"
+
+#include <cstdint>
+
+namespace pmacc::math
+{
+    /** 2 d matrix matrix definition
+     *
+     * @tparam T_Type data type used for storage of elements
+     * @tparam T_Extent pmacc::math::CT::Vector, tow dimensional vector describing extent of matrix in each dimensions
+     *
+     * @attention uses mathematical indexing
+     */
+    template<typename T_Type, typename T_Extent>
+    struct Matrix
+    {
+        PMACC_CASSERT_MSG(matrix_must_have_2_dimensional_extent, T_Extent::dim == 2);
```

Comment:
doesn't matter in execution time, since the check for correct dimension is compile time only.

Reply:
InN this case an assert is fine because only 2D is supported. Specialization would make only sense if we need the object for Sfinae calls to test which signature is available.